Validate softwire for missing PSID mapping

tsyesika opened the pull request:
This now will validate softwires which are added via `snabb config add` or `snabb config set` and ensure that every softwire has a corresponding PSID mapping and that invalid softwires are not silently accepted. It also checks that when you start the lwaftr the config file it is started with is valid. This occurs by having validation in the leader to validate softwires prior to them being sent to the data plane.

This fixes the problem described in #727.

Commit:
Add test for #727 - check softwires are matched

This adds tests to check that softwires that have no PSID map counterpart
cause an error to be raised. It tests that this occurs when adding or
setting one via snabb config and when running lwaftr bench and lwaftr run.

Commit:
Add validation for softwires using snabb config add

This adds checking for adding softwires without a PSID map counterpart
that will now produce an error when it is attempted. This is to partly
address issue #727.

Commit:
Fix pervious config test with new validation

Previously we had a test which added a softwire and read it back to
check that softwires can be added and read back correctly. This added
a softwire which didn't have a PSID map entry. This test fixes that.

Commit:
Improve tests for softwire validation

Adds error messages to explain why tests have failed so it's
easier to see what the problem is. I have also extended the
config add test to verify it not only errors but also doesn't
add the softwire.

Commit:
Fix softwire validation for config and initial start

This fixes a bug where it would error when an invalid softwire was
provided on a snabb config add but it was still added. It also adds
checking for starting the lwaftr with an invalid config.

Commit:
Add validation functions to generic leader support

Commit:
Fix config validation for non --reconfigurable too

dpino commented:
The tests are not passing. I pulled the PR and run `make test` to see what's happening. At least the soaktest is failing:

```
$ sudo ./program/lwaftr/tests/soaktest/selftest.sh
Testing: Regression test: make sure end-addr works
../data/icmp_endaddr.conf: loading compiled configuration from ../data/icmp_endaddr.o
../data/icmp_endaddr.conf: compiled configuration is up to date.
```

What's happening is that `icmp_endaddr.conf` has a PSID map configured using `end-addr`. PSID maps can defined as a range of IPv4 addresses using the `end-addr` attribute. Here's how `psid_map` is defined in `icmp_endaddr.conf`:

```
psid-map {
   addr 178.79.150.233;
   end-addr 178.79.150.236;
   psid-length 0;
   shift 16;
}
```

Which is the same as defining 4 independent psid-maps for IPv4 addresses 178.79.150.233, 178.79.150.234, 178.79.150.235 and 178.79.150.236.

Hope this shed some light.

Commit:
Lookup psid entries using rangemap for range support

This switches to maintaining an instance of the lwaftr's binding table
in the leader to allow for looking up PSID entries, specifically those
which have a range (i.e. specify an "end-addr"). It still needs to
maintain the binding table where possible to prevent it being re-build
on every validation (expensive).

xray7224 commented:
I have pushed a commit which shows my work today. This now builds the binding table in the leader which is then used to perform the lookups allowing for the range support. This does it in a smilier fashion to how the data-plane does it. The code doesn't yet maintain the binding table (planning to add it tomorrow) so it currently just rebuilds it every time it needs it which is obviously in-effeiciant. 

Let me know what you think @wingo. This is the only solution I can think of without involving the data-plane (which we obviously can't do for timing concerns).

wingo commented on src/apps/config/support/snabb-softwire-v1.lua:
```diff
@@ -12,6 +13,51 @@ local ctable = require('lib.ctable')
 local cltable = require('lib.cltable')
 local path_mod = require('lib.yang.path')
 local generic = require('apps.config.support').generic_schema_config_support
+local binding_table = require("apps.lwaftr.binding_table")
```
I think we should avoid using this module; for me it is mixing layers that shouldn't be mixed.  Instead we can just look up the IPv4 address of a softwire in the psid_map list; does that sound OK to you?

tsyesika replied:
My code use to work this way but I don't know how to get an efficient effective way to search for ipv4's which fall within a range (i.e. using `end-addr` - look at @dpino's comment above to see the problem I hit using this method). I feel as though I'd probably just end up re-implementing what `apps.lwaftr.rangemap` implements if I tried to do it without this module.

I could always use the `rangemap` directly instead of using the code in the binding table to initialise it however I didn't really see the point in using some of the code under `apps.lwaftr` and not other parts (which do what I need them to).

WDYT?

wingo commented on src/apps/config/support/snabb-softwire-v1.lua:
```diff
@@ -12,6 +13,51 @@ local ctable = require('lib.ctable')
+
+
+local binding_table_instance
+local function get_binding_table(conf)
```
For the above reason I think we can and probably should do everything on a configuration instance.

wingo commented on src/apps/config/support/snabb-softwire-v1.lua:
```diff
@@ -12,6 +13,51 @@ local ctable = require('lib.ctable')
+
+local function validate_update(config, verb, path, data)
+   -- Validate softwire has PSID Map entry.
+   if path == "/softwire-config/binding-table/softwire" then
```
Looks right to me.

wingo commented on src/program/lwaftr/setup.lua:
```diff
@@ -58,6 +63,9 @@ function lwaftr_app(c, conf)
 
+   -- Validate configuration file we've been given.
+   validate_config(conf)
+
```
I think we probably don't want to do this here.  I mean, this function will be called on any configuration change, including incremental updates, right?  We probably don't want to cause an O(n) verification when our O(1) update already verified the thing.  So when do we need to verify?  Perhaps only when loading the initial configuration from disk.

tsyesika replied:
Hm, you're right I think. I'll try and think of a better place to put it, I need a place which works for the initial validation for unreconfigurable and reconfigurable instances of the lwaftr alike.

Good catch.

wingo commented on src/apps/config/support.lua:
```diff
@@ -177,6 +177,8 @@ end
 
 generic_schema_config_support = {
+   validate_config = function() end,
+   validate_update = function () end,
```
Makes a lot of sense to me.

wingo reviewed (commented):
Looking good, some small comments.  I had kinda hoped that YANG could impose its own constraints, and that we would be able to push this validation into the model.  Maybe we can get that in the future but this PR looks like a good start for the present.

Commit:
Add snabb set test and fix get_binding_table

This adds another test which checks that erroring occurs on set also
not just add (when setting an invlid softwire). It fixes where the
initial check of the config happens to prevent it being called too
often as well as fixing how often the binding table in the leader
is rebuilt.

xray7224 commented:
I think this is ready to merge, if someone can PTAL that'd be great :)

wingo commented on src/program/lwaftr/run/run.lua:
```diff
@@ -162,6 +162,7 @@ function run(args)
    else
       setup.apply_scheduling(scheduling)
       setup_fn(c, conf, unpack(setup_args))
+      setup.validate_config(conf)
```
Probably best to move before the setup_fn call, I think

wingo reviewed (approved):
lgtm, just one maybe-nit.  thanks!

Commit:
Fix small nit with validate_config call
